Federated plugin structure

remonh87 opened the pull request:
I changed the structure a bit based on best practices from Google / Baseflow and Firebase plugins (NO ADDITIONAL FUNCTIONALITY IS CHANGED).

The structure:

- example containing the example app
- packages
     - flutter_reactive_ble => package users implements. This is now a pure dart package with business logic
     - reactive_ble_mobile => plugin for iOS and android containing the method channel implementation + protobuf   stuff.
     - reactive_ble_platform_interface => the platform interface people need to implement for other platforms

todo:

- [x]  add Melos (monorepo package management system).
- [x]  update readme
- [x]  fix CI    
- [x]  squash commits

PieterAelse reviewed (commented):
Mostly some nits, so looks good 👍 

Also tested the example app and runs/connects without issues.

remonh87 commented:
@PieterAelse  thanks for the review. I fixed all the comments can you look to it again and give your final verdict? I updated all versions to `4.0.0` since I want everything to be consistent (`reactive_ble platform_interface` v4.0.0 matches with `flutter_reactive_ble` 4.0.0. The CI is red because I still have links to the local dependencies but I ran all the commands before and everything Is in check. When you approve this I will start publishing it and the CI will be green (kind of chicken egg thingy).

PieterAelse commented on packages/reactive_ble_platform_interface/lib/src/reactive_ble_platform_interface.dart:
```diff
@@ -0,0 +1,184 @@
+
+/// The interface that implementations of `reactive_ble` must implement.
+///
+// Platform implementations should extend this class rather than implement it as
```
@remonh87 Kind reminder for this one

remonh87 replied:
ah that is on purpose. Because the first two lines are part of the DartDoc and the other is more comment for people that want to extend it

PieterAelse commented:
@remonh87 Approved with one missed comment still open, but not a biggie: https://github.com/PhilipsHue/flutter_reactive_ble/pull/338#discussion_r689360921

juliansteenbakker commented:
Looks good! Just one question, right now all platform packages a visible on pub.dev. It would be better to unlist al plaform specific packages, and just list the front-facing [flutter_reactive_ble ](https://pub.dev/packages/flutter_reactive_ble ) package. Otherwise people will try to implement platform specific packages. You can change this on the pub.dev package admin panel.

remonh87 commented:
@juliansteenbakker  thanks for the tip I fixed it directly wasn't aware this is possible :)
